packages/gatsby/src/redux/actions/public.js

```diff
@@ -614,8 +622,7 @@ const createNode = (
     node.internal = {}
   }
 
-  NODE_COUNTER++
-  node.internal.counter = NODE_COUNTER
+  node.internal.counter = getNextNodeCounter()
```

Comment:
Given that `createNode` action creator might result in `TOUCH_NODE` action and not necesairly `CREATE_NODE` - should we actually assign (and increment) node counter only when node is actually created and not touched (so somewhere in https://github.com/gatsbyjs/gatsby/blob/720119c78bcefbdba862785c3c890bfefeef5afd/packages/gatsby/src/redux/actions/public.js#L759-L790 )?

Alternatively `status` reducer might need to handle `TOUCH_NODE` action in addition to `CREATE_NODE` action to correctly keep track of last used counter?

I think I would lean toward updating counter only when `CREATE_NODE` action is dispatched for 2 reasons:
 - using `touchNode` action creator explicitely actually doesn't increment counter (so we currently end up with 2 action creators that dispatch `TOUCH_NODE` where only one of them (`createNode`) increment counter (tho this could be addressed also with adding similar code in `touchNode` - tho it doesn't make sense to me to actually do it conceptually)
 - (lesser reason, but still) it would just be _slower_ to reach `MAX_SAFE_INTEGER` if we only increment when node are actually upserted

Reply:
Good point. Not sure if it's in scope for this PR but probably makes sense to add this change anyway. Will do.